## Github Repository for
# RefSoil+: a Reference Database for Genes and Traits of Soil Plasmids
## by Taylor K. Dunivin, Jinlyung Choi, Adina Howe, and Ashley Shade

<i>This work is published.</i>


### To cite this work or code
Dunivin TK, Choi J, Howe A, Shade A. 2019. RefSoil+: a reference database for genes and traits of soil plasmids. mSystems 4:e00349-18. https://doi.org/10.1128/mSystems.00349-18.


### Abstract
Plasmids harbor transferable genes that contribute to the functional repertoire of microbial communities, yet their contributions to metagenomes are often overlooked. Environmental plasmids have the potential to spread antibiotic resistance to clinical microbial strains. In soils, high microbiome diversity and high variability in plasmid characteristics present a challenge for studying plasmids. To improve the understanding of soil plasmids, we present RefSoil+, a database containing plasmid sequences from 922 soil microorganisms. Soil plasmids were larger than other described plasmids, which is a trait associated with plasmid mobility. There was a weak relationship between chromosome size and plasmid size and no relationship between chromosome size and plasmid number, suggesting that these genomic traits are independent in soil. We used RefSoil+ to inform the distributions of antibiotic resistance genes among soil microorganisms compared to those among nonsoil microorganisms. Soil-associated plasmids, but not chromosomes, had fewer antibiotic resistance genes than other microorganisms. These data suggest that soils may offer limited opportunity for plasmid-mediated transfer of described antibiotic resistance genes. RefSoil+ can serve as a reference for the diversity, composition, and host associations of plasmid-borne functional genes in soil, a utility that will be enhanced as the database expands. Our study improves the understanding of soil plasmids and provides a resource for assessing the dynamics of the genes that they carry, especially genes conferring antibiotic resistances.

### Funding
This work was supported in part by the National Science Foundation under grants DEB number [1655425](https://www.nsf.gov/awardsearch/showAward?AWD_ID=1655425&HistoricalAwards=false) and DEB number [1749544](https://www.nsf.gov/awardsearch/showAward?AWD_ID=1749544), by the USDA National Institute of Food and Agriculture and Michigan State AgBioResearch, and by the Great Lakes Bioenergy Research Center award number [DE-SC0018409](glbrc.org) (to A.S.). Support was also provided by the Michigan State University Department of Microbiology and Molecular Genetics [Russell B. DuVall Fellowship](https://mmg.natsci.msu.edu/academics/graduate/information-for-current-graduate-students/funding/awards-fellowships/) (to T.K.D.). This work was funded in part by the DOE Center for Advanced Bioenergy and Bioproducts Innovation under award [DE-SC0018420](cabbi.bio), to J.C. and A.H.


### More info
[ShadeLab](http://ashley17061.wixsite.com/shadelab/home)
